### PR TITLE
Enable always on burger menu for responsive navigation menus.

### LINKED ISSUE #34783: [Navigation Block] Add option to use the "responsive menu" at all times
Sometimes a site may want to hide menu items behind a hamburger even on large screens. When "Enable responsive menu" is active, we should add an additional checkbox that lets users enforce that behavior. 

Example:

Video Theme 2

_I was sure there was another issue for this, but I couldn't find one. Feel free to close if I'm mistaken!_

### PR DESCRIPTION
**Pull request by vcanales**
## Description
Fixes https://github.com/WordPress/gutenberg/issues/34783

Introduces an option to keep a Navigation Menu always hidden behind a burger menu.

Classnames and text are WIP.

## How has this been tested?
- Create a Navigation Menu.
- Enable Responsiveness.
- Enable "Always on burger menu".
- Navigation Menu should be behind a Toggle Button regardless of screen size.

### Individual Switches
Kapture 2021-10-12 at 17 08 25

### Toggle Group
Kapture 2021-10-12 at 21 17 41

### CODE REVIEW

**Comment by mtias** [comment]
Cool! How far are we from being able to implement the proposed design?

image

**Comment by vcanales** [comment]
> Cool! How far are we from being able to implement the proposed design?

Closer now (https://github.com/WordPress/gutenberg/pull/35568/commits/b7587957d1fa0edaa5169ce3c05f0cd2dcb62ab8) 😉 The [new design also proposes moving these settings to the root level panel](https://github.com/WordPress/gutenberg/issues/30370) instead of being under "Display settings", but that will require a bigger refactor; I think this is a good first step.

Kapture 2021-10-12 at 21 17 41

File: packages/block-library/src/navigation/block.json [hunk 1]
```diff
@@ -36,13 +40,13 @@
 			"type": "boolean",
 			"default": true
 		},
-		"openSubmenusOnClick":{
+		"openSubmenusOnClick": {
 			"type": "boolean",
 			"default": false
 		},
-		"isResponsive": {
```
Comment (jasmussen) [inline packages/block-library/src/navigation/block.json:43]: What kind of backwards compatibility does this enable? If a theme is already using the isResponsive prop, will it be ported to the new "Mobile" setting? Is there an easy way we can make this happen?

I would love to land this one super soon if at all possible, so if there's any way we could fix this one quickly even if it means coming back to it later with a refactor, that would be great.

File: packages/block-library/src/navigation/edit.js [hunk 2]
```diff
@@ -135,7 +141,7 @@ function Navigation( {
 		className: classnames( className, {
 			[ `items-justified-${ attributes.itemsJustification }` ]: attributes.itemsJustification,
 			'is-vertical': attributes.orientation === 'vertical',
-			'is-responsive': attributes.isResponsive,
+			'is-responsive': 'never' !== attributes.overlayMenu,
```
Comment (jasmussen) [inline packages/block-library/src/navigation/edit.js:144]: In relation to the previous comment about back compat — could we maintain back compat with existing themes by keeping the property as "isResponsive", and then setting the mobile property _either_ if it's stored as `true` or `"mobile"`? 

Not necessarily as a permanent thing, we could follow up and clean things up in a followup PR — just in the name of landing this one quickly.

**Comment by jasmussen** [comment]
I pushed a small tweak to the inspector control, to make it look like this:
Screenshot 2021-10-13 at 10 13 52

I'm going to explore how we can get the Overlay Menu subheading closer.

**Comment by jasmussen** [comment]
I added a separate subheading:
Screenshot 2021-10-13 at 10 17 03

This particular subheading style is likely one we will want to revisit and unify. In this case, the change I'm making affects all `h3`s in the inspector. Any thoughts on whether that's okay, @ntsekouras? (If you have time to look, thank you)

File: packages/block-library/src/navigation/style.scss [hunk 3]
```diff
@@ -505,16 +505,18 @@
...
-			flex-direction: row;
-			width: 100%;
-			position: relative;
-			z-index: 2;
-			background-color: inherit;
-
-			.wp-block-navigation__responsive-container-close {
-				display: none;
+		&:not(.hidden-by-default) {
+			&:not(.is-menu-open) {
```
Comment (jasmussen) [inline packages/block-library/src/navigation/style.scss:509]: The rules here are a bit hard to parse, but I think they make sense. But it would be good if a themer, perhaps @kjellr could take this PR for a spin with a theme that uses the existing responsive feature, to see that it works as it should?

**Comment by jasmussen** [comment]
I tweaked the wording a bit:
Screenshot 2021-10-13 at 10 33 25

**Comment by jasmussen** [comment]
Pushed a change to simplify the two existing toggles as well, now that they have a subheading:
Screenshot 2021-10-13 at 10 45 55

**Comment by jasmussen** [comment]
Alright, I've pushed some polish to this one, tested it thoroughly, and it feels like it's in both a pretty solid place featurewise, and in my testing, it works as intended with no ill side effects.

The only missing thing is to add some back compat to existing themes that toggled on the previous `isResponsive` boolean toggle. It would be nice if we could continue to support the previous attribute, read from it and map it to the new "Mobile" state. 

Here's some test content for an existing menu that is set to "responsive":

```
<!-- wp:paragraph -->
<p>Legacy responsive menu:</p>
<!-- /wp:paragraph -->

<!-- wp:navigation {"isResponsive":true} -->
<!-- wp:page-list {"isNavigationChild":true,"showSubmenuIcon":true,"openSubmenusOnClick":false} /-->
<!-- /wp:navigation -->
```

File: packages/block-library/src/navigation/block.json (hunk 1 above)
Comment (ntsekouras) [reply packages/block-library/src/navigation/block.json:43]: We will need a deprecation to handle this properly.

Comment (jasmussen) [reply packages/block-library/src/navigation/block.json:43]: I'd be very grateful for any help here 🍦🍒

**Comment by ntsekouras** [comment]
>This particular subheading style is likely one we will want to revisit and unify. In this case, the change I'm making affects all h3s in the inspector. Any thoughts on whether that's okay, @ntsekouras?

I'm not sure if we should do this as there are labels from other controls inside `label`. I haven't tested it yet but I think not all cases will be covered..

**Comment by jasmussen** [comment]
> I'm not sure if we should do this as there are labels from other controls inside label. 

For the time being, it's not touching labels, I agree that needs a more holistic look. It only touches h3's inside the inspector — and those appear to have been mostly unused for the inspector, so it feels safe enough to me. Also CC @ciampo related to the components/labels conversation around headings we had recently. 

https://github.com/WordPress/gutenberg/pull/35568/files#diff-f6012769a71b3b8b71d12af8b6eeb62f5def7807e972f2024f44ac9a0df27e77R44

**Comment by jasmussen** [comment]
This is excellent, thank you Nik for the deprecation. I'm now seeing classic menus map seamlessly to the new:
Screenshot 2021-10-13 at 13 28 29

I'm happy to give this PR a green check, because to me this one is solid, safe, and great. But I'd love a few more eyes on this one just to increase the confidence level. Thanks everyone, and thanks Vicente for the fast work on this one!

File: packages/block-library/src/navigation/edit.js (hunk 2 above)
Comment (ntsekouras) [reply packages/block-library/src/navigation/edit.js:144]: This is handled by the deprecation

File: packages/block-library/src/navigation/edit.js [hunk 4]
```diff
@@ -271,24 +277,45 @@ function Navigation( {
...
+							/>
+							<ToggleGroupControlOption
+								value="mobile"
+								label={ __( 'Mobile' ) }
+								aria-label={ __( 'Mobile' ) }
+							/>
+							<ToggleGroupControlOption
+								value="always"
+								label={ __( 'Always' ) }
+								aria-label={ __( 'Always' ) }
```
Comment (ntsekouras) [inline packages/block-library/src/navigation/edit.js]: Nitpick: If we want the same label for `aria-label` we don't need to provide it here. If not provided it will get `label`'s value.

File: packages/block-library/src/navigation/style.scss (hunk 3 above)
Comment (kjellr) [reply packages/block-library/src/navigation/style.scss:509]: I tested this with a few existing block themes that had the responsive menu set to (what is now) "Mobile" and found no issues. @jeffikus and @scruffian are also giving it a quick spin too just to be sure but I feel pretty solid on this so far.

**Review by ntsekouras — APPROVED** [review]
Thanks for working on this Vicente! LGTM!

I'd appreciate a check for the deprecation code though by you 😄

**Comment by mtias** [comment]
Thanks for the quick work on this one!

**Comment by scruffian** [comment]
Tested on Blockbase, Quadrat, Geologist, Seedlet Blocks and Mayland Blocks. Looks good on all of them.

**Comment by vcanales** [comment]
> Thanks for working on this Vicente! LGTM!
> 
> I'd appreciate a check for the deprecation code though by you 😄

Tested, and it looks good to me too!

Thanks everyone for contributing, fast iteration, and testing! It was beautiful waking up to all this activity 🙂

**Comment by jasmussen** [comment]
💯
